### PR TITLE
SDK/Compiler - Invoke the op_transformers as early as possible

### PR DESCRIPTION
**Pull request by kvalev**
Invoking the `op_transformers` functions after the container inputs have been determined will result in an invalid argo yaml definition, in case a transformation function is used to set pod labels based on a pipeline parameter.

Before this PR:
```
  - dag:
      tasks:
      - name: cop
        template: cop
    name: parameters-in-op-transformation-functions
```

After this PR:
```
  - dag:
      tasks:
      - arguments:
          parameters:
          - name: param
            value: '{{inputs.parameters.param}}'
        name: cop
        template: cop
    inputs:
      parameters:
      - name: param
    name: parameters-in-op-transformation-functions
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1464)

### CODE REVIEW

**Comment by hongye-sun** [comment]
I don't have any problem with this PR. Alexey, do you have any comment?

**Comment by Ark-kun** [comment]
/hold
The tests are failing. I'm looking why the unit tests are not failing.

File: sdk/python/kfp/dsl/_pipeline.py [hunk 1]
```diff
@@ -128,15 +128,15 @@ class Pipeline():
   """
 
   # _default_pipeline is set when it (usually a compiler) runs "with Pipeline()"
-  _default_pipeline = None
+  _default_pipeline: 'Pipeline' = None
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/_pipeline.py]: This is not supported in Python 3.5.

File: sdk/python/kfp/dsl/_pipeline.py [hunk 2]
```diff
@@ -128,15 +128,15 @@ class Pipeline():
...
 
   @staticmethod
-  def get_default_pipeline():
+  def get_default_pipeline() -> 'Pipeline':
     """Get default pipeline. """
     return Pipeline._default_pipeline
 
   @staticmethod
-  def add_pipeline(name, description, func):
+  def add_pipeline(name: str, description: str, func) -> None:
```
Comment (Ark-kun) [inline sdk/python/kfp/dsl/_pipeline.py]: Nit: It's better to have refactorings and actual changes in separate PRs. Minimizing the PR diff is important for making the reviewing easier.

**Comment by Ark-kun** [comment]
The code is not compatible with Python 3.5 which we still support (partially due to Debian)
/approve cancel

File: sdk/python/kfp/dsl/_pipeline.py (hunk 1 above)
Comment (kvalev) [reply sdk/python/kfp/dsl/_pipeline.py]: Reverted.

File: sdk/python/kfp/dsl/_pipeline.py (hunk 2 above)
Comment (kvalev) [reply sdk/python/kfp/dsl/_pipeline.py]: Reverted.

**Comment by Ark-kun** [comment]
/lgtm
Waiting for all tests to pass.

**Comment by kvalev** [comment]
I was not aware that it should be compatible with Python 3.5 - the travis build only tests 3.6. I reverted the type hinting, would you mind taking another look @Ark-kun?

**Comment by Ark-kun** [comment]
>I was not aware that it should be compatible with Python 3.5 

In latest stable Debian the latest version of python you can get using `apt-get` is 3.5.
https://github.com/kubeflow/pipelines/blob/5061fcffcfb5e1a0eae024e2294af85cc71cb1fb/sdk/python/setup.py#L59

I'd like to get a data point from you: If the KF Pipelines SDK needed the python version that you cannot install by just running `apt-get install python3`, would that be a problem for you? Do you think that can be a problem for other potential users?

**Comment by kvalev** [comment]
Well, it would not stop me from using the SDK, but it would be somewhat annoying, if I cant just `pip install` it. Sure, you can always either compile a newer Python version yourself or run it in a docker container, but it is just not the same. I would suggest keeping the Python 3.5 compatibility - I dont think there are significant downsides.

**Comment by Ark-kun** [comment]
/approve

There seems to be a conflict. Can you please resolve it?

**Pull request merged**